border-image-* returns the initial keyword instead of the initial value when setting border-image shorthand

darinadler opened the pull request:
#### c0c6ad146d33ff23f151c5b283baedd72f840d91
<pre>
border-image-* returns the initial keyword instead of the initial value when setting border-image shorthand
<a href="https://bugs.webkit.org/show_bug.cgi?id=244657">https://bugs.webkit.org/show_bug.cgi?id=244657</a>
rdar://problem/99420050

Reviewed by Sam Weinig.

* LayoutTests/fast/borders/border-image-legacy.html: Expect initial values instead of the
keyword &quot;initial&quot; for specified style.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-image-shorthand.sub-expected.txt:
Expect PASS.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-image-shorthand.sub-expected.txt:
Ditto.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::borderImagePropertyValue const): Change the code that checks for
implicit values to not require use of the initial keyword.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeBorderImage): Set initial values using the actual values
rather than the initial keyword.

Canonical link: <a href="https://commits.webkit.org/257119@main">https://commits.webkit.org/257119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2d03e87088713a7a08dcc6674ffbd10635818d6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31064 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107369 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167647 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101835 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7570 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35893 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90417 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104006 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103536 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5692 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84510 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32766 "") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87563 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89302 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1102 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1089 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22238 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4893 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5918 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2362 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41641 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5745,11 +5753,24 @@ bool CSSPropertyParser::consumeBorderImage(CSSPropertyID property, bool importan
+            if (slice)
+                addProperty(CSSPropertyBorderImageSlice, property, slice.releaseNonNull(), important);
+            else {
+                auto value = CSSBorderImageSliceValue::create(createQuad(100, CSSUnitType::CSS_PERCENTAGE), property == CSSPropertyWebkitBorderImage);
```
Well spotted, `ApplyPropertyBorderImageModifier::applyInitialValue` doesn't set `fill` so there was this bug:

```js
var s = document.body.style;
var cs = getComputedStyle(document.body);
s.webkitBorderImage = "url()";
cs.borderImageSlice; // "100%"
s.webkitBorderImage = cs.webkitBorderImage; // "url(\"\") 100% / 1 / 0 stretch"
cs.borderImageSlice; // "100% fill"
```

Now it's consistent:

```js
s.webkitBorderImage = "url()";
cs.borderImageSlice; // "100% fill"
s.webkitBorderImage = cs.webkitBorderImage; // "url(\"\") 100% fill / 1 / 0 stretch"
cs.borderImageSlice; // "100% fill"
```

Loirooriol replied:
Further thinking about this, the serialization of the computed value will still need to be fixed anyways, and this introduces a regression:

Previously, `-webkit-border-image` had round-tripping for slice:

```js
var s = document.body.style;
s.borderImage = "url()";
[s.borderImageSlice, // "initial"
 s.webkitBorderImage += "", // "url(\"\")"
 s.borderImageSlice]; // "initial"
```
Now it serializes if the slice is implicit, even if it doesn't have fill, and when reparsing it adds fill:

```js
var s = document.body.style;
s.borderImage = "url()";
[s.borderImageSlice, // "100%"
 s.webkitBorderImage += "", // "url(\"\")"
 s.borderImageSlice]; // "100% fill"
```

So if when `-webkit-border-image` resets an omitted `border-image-slice` it adds `fill`, then the serialization code should return empty string when there is a mismatch.

So maybe just keep the old behavior of not setting `fill` when omitted.

darinadler replied:
OK, done.

Loirooriol replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=248113 for fixing the serialization of the computed `-webkit-border-image`

Loirooriol reviewed (approved):
Informal LGTM.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -5745,11 +5753,24 @@ bool CSSPropertyParser::consumeBorderImage(CSSPropertyID property, bool importan
+            if (outset)
+                addProperty(CSSPropertyBorderImageOutset, property, outset.releaseNonNull(), important);
+            else
+                addProperty(CSSPropertyBorderImageOutset, property, createQuad(0, CSSUnitType::CSS_NUMBER), important, true);
```
Why can't we keep using `addPropertyWithImplicitDefault()` here? It handles this pattern of `if (value) addExplicit() else addImplicit()`

darinadler replied:
We can, but we end up doing a lot of wasted work to create the implicit value that is thrown away whenever the explicit value is non-null. Those quads are complex objects. I don’t have a strong feeling either way.

darinadler replied:
I did it. Code looks much shorter, is less efficient.

Loirooriol replied:
Maybe `addPropertyWithImplicitDefault` should take a lambda that creates the value if needed instead of the value itself?

Loirooriol reviewed (commented):
I think you forgot to upload the new version of the patch

Loirooriol commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -30,6 +30,8 @@
 #include "CSSPropertyParser.h"
 
+#include "CSSBorderImageSliceValue.h"
+#include "CSSBorderImageWidthValue.h"
```
`consumeBorderImageComponents` is in CSSPropertyParserHelpers.cpp already, so you could move `consumeBorderImage` too and then no need to include the headers in both places.

darinadler replied:
I’d hate to delay landing this small change further for that, but I guess I could do it. At the time I started this change the file already included those headers.

Loirooriol replied:
Not blocking on this

darinadler committed:
border-image-* returns the initial keyword instead of the initial value when setting border-image shorthand
https://bugs.webkit.org/show_bug.cgi?id=244657
rdar://problem/99420050

Reviewed by Sam Weinig.

* LayoutTests/fast/borders/border-image-legacy.html: Expect initial values instead of the
keyword "initial" for specified style.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-image-shorthand.sub-expected.txt:
Expect PASS.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-image-shorthand.sub-expected.txt:
Ditto.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::borderImagePropertyValue const): Change the code that checks for
implicit values to not require use of the initial keyword.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeBorderImage): Set initial values using the actual values
rather than the initial keyword.

Canonical link: https://commits.webkit.org/257119@main
